File: java/src/main/java/linkedlist/$725_SplitLinkedListInParts.java
package linkedlist;

import common.ListNode;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class $725_SplitLinkedListInParts {
    public ListNode[] splitListToParts(ListNode root, int k) {
        ListNode[] ans=new ListNode[k];
        ListNode p=root;
        int len=0;
        while(p!=null){
            len++;
            p=p.next;
        }
        p=root;

        if(len<=k){
            for (int i = 0; i < len; i++){
                ans[i]=new ListNode(p.val);
                p=p.next;
            }
        }
        else{

            int remain =len%k;
            int precount=len/k;
            int[] counts=new int[k];
            for(int i=0;i<k;i++){
                counts[i]= remain-- >0 ? precount+1 : precount;
            }

            for(int i=0;i<k;i++){
                ans[i]=p;
                while(--counts[i] >0){
                    p=p.next;
                }
                ListNode tmp=p.next;
                p.next=null;
                p=tmp;
            }
        }

        return ans;
    }
}

File: python/linkedlist/$725_SplitLinkedListInParts.py
from typing import List

from common.ListNode import ListNode


class Solution:
    def splitListToParts(self, head: ListNode, k: int) -> List[ListNode]:
        cur, n = head, 0
        while cur:
            n += 1
            cur = cur.next
        each, remain = n // k, n % k
        cur, res, idx = head, [None] * k, 0

        while cur:
            res[idx] = cur
            last = None
            for _ in range(each + (idx < remain)):
                last = cur
                cur = cur.next
            last.next = None
            idx += 1

        return res


arr = [1, 2, 3]
_k = 5
hair = ListNode()
p = hair
for i in arr:
    p.next = ListNode(i)
    p = p.next
pr = Solution().splitListToParts(hair.next, _k)
print(pr)
